### PR TITLE
fix(flip-assist): clamp sell qty to inventory + suppress Collect duplicate fills

### PR DESCRIPTION
**Pull request by mablack01**
## Summary

Two related trading-safety fixes: FlipAssist could suggest selling more than the player actually holds, and the Collect button could cause the same physical fill to be logged twice under a new offer ID.

Part of Flip-Smart/flip-smart#975

## Problem

FlipAssist could prompt the player to sell more than they actually hold in inventory — the suggested sell quantity was sourced from an over-counted API quantity rather than being bounded by what's actually in the bag.

Separately, the in-game Collect button causes the Grand Exchange to re-detect the same physical offer under a new `offer_id`. `TransactionLogger` was treating that re-detection as a brand new fill and re-sending it, producing duplicate fill records for a single physical trade.

## Fix

1. `GrandExchangeTracker.setFocusForSell` now clamps the sell quantity to the player's actual inventory count via a new pure method, `resolveSellQuantity`, extracted specifically so it's unit-testable. It will never suggest selling more than is held.
2. `TransactionLogger` adds a churn-resistant normalized identity to its client-side seen-key guard: `rsn:itemId:isBuy:roundTrip:TYPE:cumulative`. When Collect causes a re-detection under a new `offer_id`, this normalized key still matches the prior fill and the duplicate send is suppressed at the source. This is purely additive — the idempotency key actually sent to the API is unchanged, and the round-trip lookup used to build the normalized key is a peek (non-mutating), so a suppressed duplicate can never advance the ledger state.

## Deferred work

A deeper hardening of `OfferStore`'s fill-baseline tracking (to stop re-emitting the full cumulative offer amount as if it were a fresh increment on re-detection) is intentionally deferred to a follow-up that needs in-game QA to validate safely. The paired API-side fix (Flip-Smart/flip-smart PR #978, tracked under issue #974) already neutralizes the impact of that deeper issue, so this PR's client-side suppression plus the API fix are sufficient for now.

## Tests

- `GrandExchangeTrackerSellQuantityTest` (new file) — covers the inventory clamp.
- `TransactionLoggerTest` — added a Collect-re-detection suppression case.

Full `./gradlew clean test` passes: **BUILD SUCCESSFUL**, 540 tests, 0 failures, 0 errors, 0 skipped (includes the 4 new `GrandExchangeTrackerSellQuantityTest` cases and the 14 `TransactionLoggerTest` cases, up from 13).

## QA

Plugin QA is inherently manual/in-game — Playwright cannot drive a RuneLite client, so this is expected for plugin PRs, not a gap in this one. The user will tail the RuneLite client log and manually walk the Collect path (buy partial fill → collect to bank → re-list → sell) once the paired API PR #978 is deployed.

---

### 🩺 Codacy Quality Gate
Gate green at `03f7c86` — `isUpToStandards=true`, 0 new issues.

### 🤖 Codacy AI Reviewer — deferred (in-thread rationale)
All 4 open comments deferred; none applied, per this PR's safety-critical scope on `resolveSellQuantity` and the `normalizedDedupKey`/seen-key guard (see workspace policy on Codacy fixes touching these two files).

- `TransactionLogger.java:144` — suggested adding `slot`/`totalQuantity`/`price` to `normalizedDedupKey`. Deferred: no test pins down that these fields stay stable across a Collect-driven offer re-detection, so adding them risks reintroducing the exact churn-driven duplicate-send bug this key exists to suppress.
- `TransactionLogger.java:112` — suggested stabilizing `roundTripId` for the "null → concrete" transition. Deferred: `peekRoundTripId` only returns null when `rsn` is unresolved (cycleId starts at 1 immediately once an Entry exists, so this isn't a cycle-bootstrap issue); the narrower real edge case (RSN-availability flapping across retries) has no test coverage, and any change here touches the guarded key-derivation logic directly.
- `GrandExchangeTracker.java:927` — suggested treating `inventoryCount == 0` as a hard zero ceiling instead of "unknown, trust API". Deferred: this is intentionally pinned by `usesApiQuantityWhenInventoryUnknown` (`resolveSellQuantity(3, 0) == 3`); changing it is a product/architecture decision (can RuneLite's inventory container distinguish "not yet read" from "confirmed empty"?), not a mechanical fix.
- `GrandExchangeTracker.java:935` — suggested replacing `Math.min(apiQuantity, inventoryCount)` with an inventory-first ternary. Deferred: this would let the suggested sell quantity rise to the full inventory count even when the tracked/API quantity is lower, weakening the min-bound clamp that `clampsToInventoryWhenApiOvercounts` / `keepsQuantityWhenApiAndInventoryAgree` guard against — a direct widening of the sell-quantity safety ceiling.

No code changes were made in this pass; `./gradlew test` remains 540/540.

### CODE REVIEW

File: src/main/java/com/flipsmart/trading/TransactionLogger.java [hunk 1]
```diff
@@ -93,27 +93,59 @@ private void recordPlacement(com.flipsmart.domain.offer.OfferRecord r)
...
+            return seen;
+        }
+    }
+
+    /**
+     * Identity of a logical fill/placement that survives Collect-driven offer_id and
+     * timestamp churn: {@code rsn:itemId:isBuy:roundTrip:TYPE:cumulative}. Used only
+     * for client-side send suppression; the sent idempotency key is unchanged.
+     */
+    static String normalizedDedupKey(String rsn, Integer roundTripId,
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/trading/TransactionLogger.java:148]: Acknowledged, but deferred. `normalizedDedupKey` is deliberately invariant to Collect-driven offer_id/slot/timestamp churn — that's this PR's core purpose (fix/975-fill-idempotency-collect). Neither `slot`, `totalQuantity`, nor `price` stability across a Collect re-detection is pinned down by an existing test, so adding them risks reintroducing the exact churn-driven duplicate-send bug this key exists to suppress. Deferred rather than fixed; see PR description for tracking.

File: src/main/java/com/flipsmart/trading/TransactionLogger.java [hunk 2]
```diff
@@ -93,27 +93,59 @@ private void recordPlacement(com.flipsmart.domain.offer.OfferRecord r)
...
     }
 
     private void recordFill(com.flipsmart.domain.offer.OfferRecord r, int newlyFilled, long newlySpent)
     {
         String rsn = rsnSupplier.get().orElse(null);
         String key = idempotencyKey(rsn, r, Type.FILL);
-        if (!seenKeys.add(key))
+        // Peek (non-mutating) the round trip this fill belongs to for the dedup key
+        // BEFORE the guard, so a suppressed duplicate never advances the ledger.
+        Integer roundTripId = roundTripLedger.peekRoundTripId(rsn, r.getItemId());
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/trading/TransactionLogger.java:112]: Acknowledged, but deferred. `peekRoundTripId` only returns null when `rsn` itself is unresolved — `cycleId` starts at 1 immediately once an Entry exists, so this isn't a cycle-bootstrap null. The narrower real edge case (RSN-availability flapping across retries of the same logical event) isn't covered by an existing test, and any change here touches the guarded key-derivation logic directly. Deferred rather than fixed without dedicated test coverage first.

File: src/main/java/com/flipsmart/GrandExchangeTracker.java [hunk 3]
```diff
@@ -916,6 +916,25 @@ private ActiveFlip findActiveFlipForItem(List<ActiveFlip> flips, int itemId)
...
 
+	/**
+	 * Resolve the sell quantity to prompt. Inventory is a hard ceiling — the
+	 * player can never sell more than they hold — so an over-counted API value
+	 * is clamped down to inventory. When inventory is unknown (0) we trust the
+	 * API; when the API has no tracked position we fall back to inventory.
+	 */
+	static int resolveSellQuantity(int apiQuantity, int inventoryCount)
+	{
+		if (inventoryCount <= 0)
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/GrandExchangeTracker.java:927]: Acknowledged, but deferred. The current `inventoryCount <= 0` => trust-API behavior is intentionally pinned by `usesApiQuantityWhenInventoryUnknown` (`resolveSellQuantity(3, 0) == 3`). Whether RuneLite's inventory container can distinguish 'not yet read' from 'confirmed empty' is a product/architecture question, not a mechanical fix — changing it would silently alter this safety clamp's semantics. Deferred rather than fixed.

File: src/main/java/com/flipsmart/GrandExchangeTracker.java [hunk 4]
```diff
@@ -916,6 +916,25 @@ private ActiveFlip findActiveFlipForItem(List<ActiveFlip> flips, int itemId)
...
+		{
+			return apiQuantity;
+		}
+		if (apiQuantity <= 0)
+		{
+			return inventoryCount;
+		}
+		return Math.min(apiQuantity, inventoryCount);
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/GrandExchangeTracker.java:935]: Acknowledged, but deferred. This touches the sell-quantity hard ceiling directly. Replacing `Math.min(apiQuantity, inventoryCount)` with an inventory-first ternary would let the suggested sell quantity rise to the full inventory count even when the tracked/API quantity is lower — weakening the min-bound clamp `clampsToInventoryWhenApiOvercounts` and `keepsQuantityWhenApiAndInventoryAgree` guard against. Deferred rather than fixed without a deliberate product decision on which value should win.

File: src/main/java/com/flipsmart/trading/TransactionLogger.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/trading/TransactionLogger.java:148]: Addressed in e5caf89 — the normalized dedup key now includes the GE slot (`rsn:itemId:isBuy:slot:roundTrip:TYPE:cumulative`). Concurrent same-item offers occupy different slots so they no longer collide; a real re-report keeps its slot even when offer_id churns, so same-slot suppression still holds. Verified in prod that duplicates/re-detections stay same-slot (Cow slot 7, Ray slot 6, water-orb slot 5). Mirrored on the API side (PR #978) which now scopes its watermark dedup by (round_trip, ge_slot). New test: `concurrentSameItemOffersInDifferentSlotsBothRecord`.
